client/src/main/java/org/eclipse/hono/client/impl/DelegatedCommandSenderImpl.java

```diff
@@ -0,0 +1,138 @@
+
+package org.eclipse.hono.client.impl;
+
+import java.net.HttpURLConnection;
+import java.util.Objects;
+
+import org.apache.qpid.proton.message.Message;
+import org.eclipse.hono.client.DelegatedCommandSender;
+import org.eclipse.hono.client.HonoConnection;
+import org.eclipse.hono.client.ServerErrorException;
+import org.eclipse.hono.client.ServiceInvocationException;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.CommandConstants;
+import org.eclipse.hono.util.MessageHelper;
+
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.tag.Tags;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.proton.ProtonDelivery;
+import io.vertx.proton.ProtonQoS;
+import io.vertx.proton.ProtonSender;
+
+/**
+ * A Vertx-Proton based sender to send command messages to the downstream peer as part of delegating them to be
+ * processed by another command consumer.
+ */
+public class DelegatedCommandSenderImpl extends AbstractSender implements DelegatedCommandSender {
+
+    DelegatedCommandSenderImpl(
+            final HonoConnection connection,
+            final ProtonSender sender) {
+
+        super(connection, sender, "", "");
+    }
+
+    @Override
+    protected String getTo(final String deviceId) {
+        return null;
+    }
+
+    @Override
+    public String getEndpoint() {
+        return CommandConstants.COMMAND_ENDPOINT;
+    }
+
+    @Override
+    protected Future<ProtonDelivery> sendMessage(final Message message, final Span currentSpan) {
+        return runSendAndWaitForOutcomeOnContext(message, currentSpan);
+    }
+
+    @Override
+    public Future<ProtonDelivery> sendAndWaitForOutcome(final Message message) {
+        return sendAndWaitForOutcome(message, null);
+    }
+
+    @Override
+    public Future<ProtonDelivery> sendAndWaitForOutcome(final Message rawMessage, final SpanContext parent) {
+        Objects.requireNonNull(rawMessage);
+
+        final Span span = startSpan(parent, rawMessage);
+        span.setTag(MessageHelper.APP_PROPERTY_TENANT_ID, MessageHelper.getTenantId(rawMessage));
+        span.setTag(MessageHelper.APP_PROPERTY_DEVICE_ID, MessageHelper.getDeviceId(rawMessage));
+        TracingHelper.injectSpanContext(connection.getTracer(), span.context(), rawMessage);
+
+        return runSendAndWaitForOutcomeOnContext(rawMessage, span);
+    }
+
+    private Future<ProtonDelivery> runSendAndWaitForOutcomeOnContext(final Message rawMessage, final Span span) {
+        return connection.executeOrRunOnContext(result -> {
+            if (sender.sendQueueFull()) {
+                final ServiceInvocationException e = new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "no credit available");
+                logError(span, e);
+                span.finish();
+                result.fail(e);
+            } else {
+                sendMessageAndWaitForOutcome(rawMessage, span).setHandler(result.completer());
```

Comment:
`Future.completer()` will be removed in vert.x 4 so I wouldn't use it anymore ...

Reply:
Ok, I've changed that.